restapi/api/reservation/reserve_endpoints.go

```diff
@@ -0,0 +1,138 @@
+			diff := currentTimestamp - reservedDevice.LastUsedTimestamp
+
+			if diff > 300000 {
+				delete(reservedDevicesMap, udid)
+			}
+		}
+		reserveMutex.Unlock()
+	}
+}
+
+// Reserve device access
+// List          godoc
+// @Summary      Reserve a device
+// @Description  Reserve a device by provided UDID
+// @Tags         reserve
+// @Produce      json
+// @Success      200  {object} reservedDevice
+// @Router       /reserve/:udid [post]
+func ReserveDevice(c *gin.Context) {
+	udid := c.Param("udid")
+	reservationID := randomReservationID()
+
+	reserveMutex.Lock()
+	defer reserveMutex.Unlock()
+
+	// Check if there is a reserved device for the respective UDID
+	device := reservedDevicesMap[udid]
+	if device == nil {
+		newReservedDevice := reservedDevice{ReservationID: reservationID, LastUsedTimestamp: time.Now().UnixMilli()}
+		reservedDevicesMap[udid] = &newReservedDevice
+	} else {
+		c.IndentedJSON(http.StatusOK, reservedDevice{Message: "Already reserved"})
+		return
+	}
+
+	c.IndentedJSON(http.StatusOK, reservedDevice{ReservationID: reservationID})
+}
+
+// Release device access
+// List          godoc
+// @Summary      Release a device
+// @Description  Release a device by provided UDID
+// @Tags         reserve
+// @Produce      json
+// @Success      200  {object} reservedDevice
+// @Failure      404  {object} reservedDevice
+// @Router       /reserve/:udid [delete]
+func ReleaseDevice(c *gin.Context) {
+	udid := c.Param("udid")
+
+	reserveMutex.Lock()
+	defer reserveMutex.Unlock()
+
+	// Check if there is a reserved device for the respective UDID
+	device := reservedDevicesMap[udid]
+	if device == nil {
+		c.IndentedJSON(http.StatusNotFound, reservedDevice{Message: "Not reserved"})
+		return
+	} else {
+		delete(reservedDevicesMap, udid)
+		c.IndentedJSON(http.StatusOK, reservedDevice{Message: "Successfully released"})
+	}
+}
+
+// Get all reserved devices
+// List          godoc
+// @Summary      Get a list of reserved devices
+// @Description  Get a list of reserved devices with UDID, ReservationID and last used timestamp
+// @Tags         reserve
+// @Produce      json
+// @Success      200  {object} []reservedDevice
+// @Router       /reserved-devices [get]
+func GetReservedDevices(c *gin.Context) {
+	reserveMutex.Lock()
+	defer reserveMutex.Unlock()
+
+	var reserved_devices []reservedDevice
+	if len(reservedDevicesMap) == 0 {
+		c.IndentedJSON(http.StatusOK, reservedDevice{Message: "No reserved devices found"})
+		return
+	} else {
+		// Build the JSON array of currently reserved devices
+		for udid, device := range reservedDevicesMap {
+			reserved_devices = append(reserved_devices, reservedDevice{
+				UDID:              udid,
+				ReservationID:     device.ReservationID,
+				LastUsedTimestamp: device.LastUsedTimestamp,
+			})
+		}
+	}
+
+	c.IndentedJSON(http.StatusOK, reserved_devices)
+}
+
+const charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+
+func randomReservationID() string {
```

Comment:
better to use UUIDs generated by a library like https://github.com/google/uuid
for this

Reply:
good point, done